ngx_http_uploadprogress_module.c

```diff
@@ -293,65 +339,161 @@ get_tracking_id(ngx_http_request_t * r)
             i = 0;
         }
 
-        if (header[i].key.len == upcf->header.len
-            && ngx_strncasecmp(header[i].key.data, upcf->header.data,
-                           header[i].key.len) == 0) {
+        if (
+            header[i].key.len == upcf->header.len &&
+            ngx_strncasecmp(header[i].key.data, upcf->header.data, header[i].key.len) == 0
+        ) {
             ret = ngx_calloc(sizeof(ngx_str_t), r->connection->log );
             ret->data = header[i].value.data;
             ret->len = header[i].value.len;
-            ngx_log_debug1(NGX_LOG_DEBUG_HTTP, r->connection->log, 0, 
-                            "upload-progress: get_tracking_id found header: %V", ret);
+            ngx_log_debug1(NGX_LOG_DEBUG_HTTP, r->connection->log, 0,
+                "upload-progress: get_tracking_id found header: %V", ret);
             return ret;
         }
     }
 
-    ngx_log_debug0(NGX_LOG_DEBUG_HTTP, r->connection->log, 0, 
-                    "upload-progress: get_tracking_id no header found");
+    ngx_log_debug0(NGX_LOG_DEBUG_HTTP, r->connection->log, 0,
+        "upload-progress: get_tracking_id no header found");
 
     /* not found, check as a request arg */
     /* it is possible the request args have not been yet created (or already released) */
     /* so let's try harder first from the request line */
-    args.len =  r->args.len;
-    args.data = r->args.data;
-    
+    args = r->args;
+
     if (args.len && args.data) {
-        ngx_log_debug0(NGX_LOG_DEBUG_HTTP, r->connection->log, 0, 
-                       "upload-progress: get_tracking_id no header found, args found");
+        ngx_log_debug0(NGX_LOG_DEBUG_HTTP, r->connection->log, 0,
+            "upload-progress: get_tracking_id no header found, args found");
         i = 0;
         p = args.data;
         do {
             ngx_uint_t len = args.len - (p - args.data);
-            if (len >= (upcf->header.len + 1) && ngx_strncasecmp(p, upcf->header.data, upcf->header.len) == 0
-                && p[upcf->header.len] == '=') {
-              ngx_log_debug1(NGX_LOG_DEBUG_HTTP, r->connection->log, 0, 
-                             "upload-progress: get_tracking_id found args: %s",p);
+            if (
+                len >= (upcf->header.len + 1) &&
+                ngx_strncasecmp(p, upcf->header.data, upcf->header.len) == 0 &&
+                p[upcf->header.len] == '='
+            ) {
+                ngx_log_debug1(NGX_LOG_DEBUG_HTTP, r->connection->log, 0,
+                    "upload-progress: get_tracking_id found args: %s",p);
                 i = 1;
                 break;
             }
-            if (len<=0)
+            else if (!len) {
                 break;
-        } 
-        while(p++);
+            }
+        } while (p++);
```

Comment:
When I went through the source for formatting, this loop condition (and the place in the duplicated routine) irritated me somewhat, as `p` is guaranteed non-NULL. May `*p++` be the actually intended check?

Reply:
I don't think`args.data` is null terminated anyway, so `*p++` is not even a good guard here.

I think the intent is that `len` is going to be 0 as we scan the args.data and reach it's length, which will make the loop exit with the `else` condition.

So basically we never check anything in that while condition.

We could change this to a `for` loop to make this more correct.

Reply:
A first correction (as part of this PR) might be as follows:

```diff
diff --git a/ngx_http_uploadprogress_module.c b/ngx_http_uploadprogress_module.c
index 44a7852a..ca899da9 100644
--- a/ngx_http_uploadprogress_module.c
+++ b/ngx_http_uploadprogress_module.c
@@ -382,10 +382,7 @@ get_tracking_id(ngx_http_request_t * r)
                 i = 1;
                 break;
             }
-            else if (!len) {
-                break;
-            }
-        } while (p++);
+        } while (len > 0);
 
         if (i) {
             start_p = p += upcf->header.len + 1;
@@ -475,10 +472,7 @@ get_tracking_ids_mul(ngx_http_request_t * r)
                 i = 1;
                 break;
             }
-            else if (!len) {
-                break;
-            }
-        } while (p++);
+        } while (len > 0);
 
         if (i) {
             start_p = p += upcf->header_mul.len + 1;
```

That's not perfect either, but already an improvement over the initial variant. The overall routine looks like it has some unconventional way to parse things, so a proper refactoring is needed for cleaning up things. Like that `i` being used as a bool; instead of using `bool` for this sort of thing.

Reply:
Yes, that would work better indeed.